### PR TITLE
Fix compiling under Windows

### LINKED ISSUE #17: convert() fails under windows
the convert() function creates a CompiledClassifier() class object, that calls build_classifier() in its initialisation. It does not pass a compiler parameter, so unless the build machine has gcc installed it will always fail at this point?

Im going to fix this in my forked repo by back-propogating 'compiler' up to convert(), but if you have a better alternative i would be happy to create the PR to fix it.

Cheers,Mark

### PR DESCRIPTION
**Pull request by micooke**
Fixes #17 

I created a mini-release 0.9.1 (setup.py), but i understand that this may not suit your workflow

### CODE REVIEW

**Comment by jonnor** [comment]
Hi @micooke, thanks for the pull request.
Have you tested this with a compiler on Windows? If so, which one? I feared that more things than just the executable needed to change to work with MSVC.

Having to pass the parameters down through the entire chain is a bit annoying, but it may be the best option.
An alternative (albeit quite Unixy in style) would maybe be to respect the `CC` environment variable?

**Comment by micooke** [comment]
My appologies, i seem to have jumped the gun a bit. I rushed a patch for another developer and haven't fully tested it yet.
I also noticed that a test failed, which i didn't notice. I initially passed the compiler through as an arg, but realised i couldn't do that without adding an exception to ```run_window_function```.

I agree, passing the compiler down is nasty. Ill keep playing with it for a bit if you are happy to wait.

**Comment by jonnor** [comment]
No worries, happy to see updates coming into a PR after the initial creation :)

**Comment by micooke** [comment]
@jonnor - Yep it fails under windows regardless.
Im looking at using distutils.ccompiler, which would scrap the compiler parameter and use the same compiler that setup.py uses. 
For reference; under windows, distutils on [py3.5+ requires MSVC++ 14.X](https://wiki.python.org/moin/WindowsCompilers#Which_Microsoft_Visual_C.2B-.2B-_compiler_to_use_with_a_specific_Python_version_.3F).

I'm currently working through compiler options as currently eml_test.h [line 49 and line 51](https://github.com/emlearn/emlearn/blob/master/emlearn/eml_test.h#L48-L51) are throwing errors [C2057](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2057?view=vs-2019),C2466,C2133 as i assume its being compiled without C99 opts.

In other words, i should have a non-hacky solution fairly soon.
As a side note : did you want __version__ in the emlearn package so that you can do i.e.
```python
import emlearn
print(emlearn.__version__) # "0.9.0"
```
I would probably put it in ```__init__.py``` and reference it in ```setup.py``` ¯\\\_(ツ)\_/¯

**Comment by micooke** [comment]
Yeah the issue is the Variable Array Length vars. The only way I've found to fix this without changing the code is to compile the c code as cpp under MSVC. One problem with that (apart from it no longer being compiled as c, hence name mangling etc) is that i dont think i can do that with distutils. AFAIK, MSVC doesn't support full C99 in any variant.

The other way would be to use  ```#define``` in the test code instead of the ```const int32_t```. Considering this is a header, its a little nasty. But its the only way forward from what i can tell.

**Comment by jonnor** [comment]
Ah, classic with MSVC having poor C99 support. But emlearn supports being compiled as C++, so I would consider that acceptable. In fact that is what the Python bindings do (with pybind11). That has some build in `setup.py` (setuptools.command.build_ex) that I **believe** also works on Windows? Perhaps that can be used instead of distutils?

But yes, for arrays that are in practice constants I am fine with using `#define` instead. Can be separate patch/PR, should be easy to review and accept.

**Comment by micooke** [comment]
No problems, ill get on to that.
I was just looking at that file and i did notice another thing : shouldn't [this](https://github.com/emlearn/emlearn/blob/master/emlearn/eml_test.h#L14) return an EmlError instead of an int? I know, functionally the same for this.

And thoughts on adding ```__version__```?

Just to clarify that my intent is to separate this in to different PRs:
This PR : update to compilation/s
PR2 : C99 VAL update for MSVC compatibility
PR3 ? : __version__ - only if you wanted this (i find it quite handy when reviewing my module versions)

**Comment by jonnor** [comment]
Yep, that should indeed be EmlError, good catch.

No objection to adding __version__ attribute

The PR plan sounds great!

**Comment by micooke** [comment]
I will keep on doing some internal testing before my next push.
I found another C99 issue in ```test/bench.c``` and missed adding the math library for ```test/test_benchmark.py```

Ill test the full gamut on msvc and gcc before my next push.
Sorry for the erroneous pushes, I had configuration issues with getting pytest running locally so hadn't been able to test fully until now.

**Comment by micooke** [comment]
I believe i have resolved these issues, expect an update shortly.
Funnily enough, i made some changes and it worked for MSVC but not gcc (mismatch between python and compiled c-code). It took me a lot of head scratching to isolate this, but i realised that it happened with successive compile & links; the second c&l was linking to the old compiled code. This was easily fixed by adding a ```force=1``` var to the ```new_compiler()``` object.

**Comment by jonnor** [comment]
The compiler parameter is now unused right, after the PRs that changed it to use distutils? So do we this PR at all? If there still is something, can you reduce it to just the needed changes and rebase on master?

**Comment by micooke** [comment]
Yeah no more compiler param. I reverted the changes I made and renamed the PR, but happy to create a new one with basically a diff of the last commit and the current master

**Comment by micooke** [comment]
You should be able to cherry pick just the changes you need, but I've never done that before

**Comment by jonnor** [comment]
Cherry-picking cannot be done through Github unfortunately. I also don't have a Windows machine available for testing right now, so I am a bit hesitant to do the cherry picking myself. So could you make a new PR against master with just the needed?

After we get that in, I guess we are ready for a new release.

**Comment by micooke** [comment]
Will do, ill try and get that to you by this time tomorrow

**Comment by micooke** [comment]
You will laugh, but before i realised I needed the force=1 i was like "yeah i fixed it for Windows".. Check it on linux again and its broken :scream:

**Comment by jonnor** [comment]
Superseeded by #24
